### PR TITLE
Rewrite readme for hugging face transformers integration

### PR DESCRIPTION
**Pull request by markurtz**

### CODE REVIEW

File: integrations/huggingface-transformers/README.md [hunk 1]
```diff
@@ -13,423 +13,106 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-python run_glue.py \
-  --student_model_name_or_path bert-base-cased \
-  --task_name MNLI \
+This integration adds a `--recipe` argument to the [`run_qa.py` script](https://github.com/neuralmagic/yolov3/blob/master/train.py) among others.
+The argument loads an appropriate recipe while preserving the rest of the training pipeline.
+Popular recipes used with this argument are found in the [`recipes` folder](./recipes).
+Distillation arguments to support student-teacher distillation are additionally added to the scripts as they help improve the recovery while sparsifying.
+Otherwise, all other arguments and functionality remain the same as the original repository.
+
+For example, pruning and quantizing a model on the SQAD dataset can be done by running the following command from within the root of this integration's folder:
```
Comment (jeanniefinks) [inline integrations/huggingface-transformers/README.md:62]:
```suggestion
For example, pruning and quantizing a model on the SQuAD dataset can be done by running the following command from within the root of this integration's folder:
```

File: integrations/huggingface-transformers/README.md [hunk 2]
```diff
@@ -13,423 +13,106 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-|-----------------------|----------	|---------|---------|------------|--------------|----------|----------|
-| bert-base-uncased 	|0        	|no       |no       |2           |0             |88.32442  |81.10690  |
-| bert-base-uncased 	|80        	|no       |no       |30          |18            |84.06276  |74.63576  |
-| bert-base-uncased 	|90        	|no       |no       |30          |18            |79.64549  |68.50520  |
-| bert-base-uncased 	|97       	|no       |no       |30          |18            |70.42570  |57.29423  |
-| bert-base-uncased 	|0        	|yes      |no       |2           |0             |89.02277  |82.03406  |
-| bert-base-uncased 	|80        	|yes      |yes      |30          |18            |88.03192  |80.81362  |
-| bert-base-uncased 	|90        	|yes      |yes      |30          |18            |85.63751  |77.41721  |
-| bert-base-uncased 	|97       	|yes      |yes      |30          |18            |75.01276  |63.94513  |
+The setup_integraiton.sh file will clone the transformers repository with the SparseML integration as a subfolder.
```
Comment (jeanniefinks) [inline integrations/huggingface-transformers/README.md:45]:
```suggestion
The `setup_integraiton.sh` file will clone the transformers repository with the SparseML integration as a subfolder.
```

File: integrations/huggingface-transformers/README.md [hunk 3]
```diff
@@ -13,423 +13,106 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+After the repo has succesfully cloned, transformers and datasets will be installed along with any necessary dependencies.
```
Comment (jeanniefinks) [inline integrations/huggingface-transformers/README.md:46]:
```suggestion
After the repo has successfully cloned, transformers and datasets will be installed along with any necessary dependencies.
```

File: integrations/huggingface-transformers/README.md [hunk 4]
```diff
@@ -13,423 +13,106 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-        "betas": (args.adam_beta1, args.adam_beta2),
-        "eps": args.adam_epsilon,
-    }
-    optimizer_kwargs["lr"] = args.learning_rate
-    return optimizer_cls(optimizer_grouped_parameters, **optimizer_kwargs)
+| Folder/File Name     | Description                                                                                                           |
+|----------------------|-----------------------------------------------------------------------------------------------------------------------|
+| recipes              | Typical recipes for sparsifying NLP models along with any downloaded recipes from the SparseZoo.                      |
+| tutorials            | Tutorial walkthroughs for how to sparsify NLP models using recipes.                                                   |
+| transformers         | Integration repository folder used to train and sparsify NLP models (setup_integration.sh must run first).            |
```
Comment (jeanniefinks) [inline integrations/huggingface-transformers/README.md:94]:
```suggestion
| transformers         | Integration repository folder used to train and sparsify NLP models (`setup_integration.sh` must run first).            |
```

File: integrations/huggingface-transformers/README.md [hunk 5]
```diff
@@ -13,423 +13,106 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-        unique_id += 1
-        # extra is used as additional data but sparseml doesn't support it
-    return (
-        torch.from_numpy(np.array([np.array(input_ids, dtype=np.int64)])),
-        torch.from_numpy(np.array([np.array(input_mask, dtype=np.int64)])),
-        torch.from_numpy(np.array([np.array(segment_ids, dtype=np.int64)])),
-    )
+### Exporting for Inference
 
+After sparsifying a model, the run_qa.py script can be run with the `--onnx_export_path` argument to convert the model into an [ONNX](https://onnx.ai/) deployment format.
```
Comment (jeanniefinks) [inline integrations/huggingface-transformers/README.md:100]:
```suggestion
After sparsifying a model, the `run_qa.py` script can be run with the `--onnx_export_path` argument to convert the model into an [ONNX](https://onnx.ai/) deployment format.
```

File: integrations/huggingface-transformers/README.md [hunk 6]
```diff
@@ -13,423 +13,106 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-```
+This directory combines the SparseML recipe-driven approach with the
+[huggingface/transformers](https://github.com/huggingface/transformers) repository.
+By integrating the robust training flows in the `transformers` repository with the SparseML code base,
+we enable model sparsification techniques on popular NLP models such as [BERT](https://arxiv.org/abs/1810.04805)
+creating smaller and faster deployable versions.
+The techniques include, but are not limted to:
+- Pruning
+- Quantization
+- Pruning + Quantization
```
Comment (jeanniefinks) [inline integrations/huggingface-transformers/README.md:27]:
```suggestion
- Pruning and Quantization
```

**Review by jeanniefinks — CHANGES_REQUESTED** [review]
Few no brainer comments, thanks.

File: integrations/huggingface-transformers/README.md [hunk 7]
```diff
@@ -13,423 +13,106 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+  --do_eval \
+  --per_device_eval_batch_size 64 \
+  --max_seq_length 384 \
+  --doc_stride 128 \
+  --cache_dir cache \
+  --preprocessing_num_workers 6 \
+  --onnx_export_path MODELS_DIR/bert-base-12layers_prune80/onnx
 ```
+
+The DeepSparse Engine accepts ONNX formats and is engineered to significantly speed up inference on CPUs for the sparsified models from this integration.
```
Comment (jeanniefinks) [inline integrations/huggingface-transformers/README.md:117]:
```suggestion
The DeepSparse Engine [accepts ONNX formats](https://docs.neuralmagic.com/sparseml/source/onnx_export.html) and is engineered to significantly speed up inference on CPUs for the sparsified models from this integration.
```

**Review by jeanniefinks — DISMISSED** [review]

**Review by bfineran — DISMISSED** [review]

File: integrations/huggingface-transformers/README.md [hunk 8]
```diff
@@ -13,423 +13,106 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-)
-exporter.export_onnx(sample_batch=sample_batch)
+For example, the following command can be run from within the integration's folder to export a trained/sparsified model's checkpoint:
+```bash
+python transformers/examples/pytorch/question-answering/run_qa.py \
+  --model_name_or_path MODELS_DIR/bert-base-12layers_prune80 \
+  --dataset_name squad \
+  --do_eval \
+  --per_device_eval_batch_size 64 \
+  --max_seq_length 384 \
```
Comment (natuan) [inline integrations/huggingface-transformers/README.md:110]: Can you help add this option: --output_dir MODELS_DIR/bert-base-12layers_prune80/eval

File: integrations/huggingface-transformers/README.md [hunk 9]
```diff
@@ -13,423 +13,106 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+Recipes encode the instructions and hyperparameters for sparsifying a model using modifiers to the training process.
+The modifiers can range from pruning and quantization to learning rate and weight decay.
+When appropriately combined, it becomes possible to create highly sparse and accurate models.
 
-Training without distillation for MNLI where model is pruned to 80% sparsity. To add distillation just include the command ```sh --teacher_model_name_or_path <your teacher model>```
-```sh
-python run_glue.py \
-  --student_model_name_or_path bert-base-cased \
-  --task_name MNLI \
+This integration adds a `--recipe` argument to the [`run_qa.py` script](https://github.com/neuralmagic/yolov3/blob/master/train.py) among others.
```
Comment (natuan) [inline integrations/huggingface-transformers/README.md:56]: Correct link: https://github.com/neuralmagic/transformers/blob/master/examples/pytorch/question-answering/run_qa.py

**Review by natuan — CHANGES_REQUESTED** [review]

**Review by bfineran — APPROVED** [review]

**Review by natuan — APPROVED** [review]
